### ISSUE TITLE
CBE: Online Rekeying

### ISSUE THREAD

**Issue by m-stein** [cbe-project/cbe#51]
It shall be possible to instruct the CBE to get rid of the currently used symmetric block encryption key completely and replace it with a new one. This will be a presumably time-intensive operation for commonly dimensioned block devices and shall therefore be executed online. I.e., the CBE keeps processing certain requests from the client while re-keying is in progress. Those are reading, writing and synchronizing blocks, creating a plaintext backup and discarding snapshots.

For the sake of simplicity, other operations like starting further re-keying operations, creating encrypted backups and creating snapshots are not processed in parallel to the re-keying in progress. They are instead deferred until the re-keying in progress has finished.

A re-keying process is split up into a series of many small atomic operations. If the CBE device is turned off and on again during a re-keying, the CBE automatically continues re-keying from the point reached with the last finished atomic operation.

**Comment by m-stein** [comment]
My current approach to this looks as follows (part 1):

__CBE interface__

Re-keying is a new operation type in the existing request structure that is used to communicate requests to the CBE. So, users can control re-keying operations via the existing CBE interface. To the user, a rekeying operation is done as soon as the corresponding request was returned as successful.

__Request Pool__

Inside the CBE, a re-keying request is submitted to the Request-Pool module. This module adds the request to the tail of its request queue and continues processing earlier requests until of them all those that must be finished before re-keying are finished. The process of re-keying a CBE device actually consists of a series of smaller atomic operations and after each atomic operation, the CBE is in a state where other requests might be processed before continuing with the re-keying.

The state of the re-keying must be reflected through the state of the CBE Superblock. Therefore, the Request Pool actually doesn't need to know about the progress of the re-keying or even the type of atomic operation needed next. It merely instructs a new Superblock Control module to execute the next atomic operation of the re-keying and to return afterwards whether it was the last one or not. This way, the Request Pool is able to schedule other allowed requests between the atomic re-keying operations and to determine when to tell the client that his re-keying request is finished.

__Superblock Control__

The new Superblock-Control module is currently only used for re-keying but shall later provide all Superblock-related operations through a typical CBE-module interface. When it receives a primitive to do one atomic step of re-keying from the Request-Pool, the Superblock-Control module first checks in which stage the re-keying is by reading the new state variable from the Superblock. There are 3 stages:

1. No re-keying in progress (initialize re-keying)
2. VBD re-keying (re-key one VBA at a time)
3. Free Tree re-keying and clean up (re-key and clean up one Free Tree branch at a time)    

In stage 1, the Superblock-Control module will first ask the new Trust-Anchor module to generate a new symmetric key that it will then store plaintext in its Superblock copy next to the old symmetric key with a key ID that is one greater than the old one. With the new key in place, it can set its Superblock copy to stage 2 and set the new Superblock.Rekeying_Curr_VBA member to 0. Next, it will ask the Cache module to synchronize all metadata blocks. Then, the Trust-Anchor module is requested to encrypt the new symmetric key, so, in the next step, the Block-IO module can be asked to write back the Superblock copy. Then, the Block IO is requested to synchronize all Blocks. To finish stage 1, the new Superblock hash is stored in the new Trust-Anchor module. Then, the Request Pool receives a success for its primitive.

In stage 2, the Superblock-Control module just tells the new VBD-Rekeying module to re-key one VBA, more precisely, the VBA that is Superblock.Rekeying_Curr_VBA. This is done for all active Snapshots, thus the snapshots must be handed over to the VBD-Rekeying module together with the old and the new symmetric key ID. Once the VBD Rekeying returns a success it also returns the updated roots of all active snapshots. These are then updated in the Superblock copy, as well as the Superblock.Rekeying_Curr_VBA. If it was the last VBA, the Superblock is moved to stage 3 and Superblock.Rekeying_Curr_VBA is set to 0 again. the Cache gets flushed, the new Superblock written back, Block IO synchronized, and the new Superblock Hash secured at the Trust Anchor. Then, the Request Pool receives a success.

Stage 3 is pretty similar to stage 2, except that each time the new Free-Tree-Rekeying module is asked to re-key and clean-up the one branch that is identified through Superblock.Rekeying_Curr_VBA instead of the VBD-Rekeying. As soon as the last Free Tree branch was handled this way, the old symmetric key can be removed from the Superblock and, again, all is synchronized and written back, the Superblock hash secured. Then the Superblock Control returns a hint to the Request Pool, that this was the last step of re-keying.

**Comment by m-stein** [comment]
There is already some progress made regarding this issue, which can be inspected on my m-stein/cbe/cbe_rekeying branch together with the cnuke/genode/cbe_wip branch.

**Comment by m-stein** [comment]
The rekeying already works fine for a great amount of VBAs. Unfortunately, I had to discover yesterday that it leaks blocks in the free tree. I.e., blocks are left as "reserved" in the free tree and will never be freed again although they could.

I think this is actually a problem of the original design: 

_Some background_

The problem occurs only with inner nodes of the VBD tree of an outdated VBD version (i.e. the blocks are not part of the most recent version of the virtual block device anymore). If one of these inner nodes is updated during rekeying, CoW must be applied and therefore the free tree is asked to allocate a new PBA. The free tree finds a free PBA but because it is meant to be part of an outdated VBD version, there are two things that are different from "normal" allocations.

First, the original PBA - the one that currently references the block that shall be updated - also has an entry in the free tree and it's marked "reserved" because this is the case for all blocks of outdated VBD versions (note, that we don't have any means to find the free tree entry of the original PBA efficiently). Second, the free tree entry of the new PBA must stay in the free tree because it will be used for a block of an outdated VBD version.

So, there will be no exchange of free tree entries in this case but merely an update of the entry of the new PBA to state that it is now "reserved". The entry of the original PBA, however, cannot be updated efficiently. Nonetheless it would be fine to re-use it as soon as we are done rekeying the current VBA. Therefore, for each free tree entry, we remember the last VBA it was used for and also its last key ID. Because the original PBA of the above example is still marked with the old key ID we can check whether its VBA was already rekeyed (we know this) and, if so, ignore its "reserved" state, i.e. recognize it as "free".

_The problem_

This works just fine the first time an inner node of an outdated VBD version is updated during rekeying. Because then, the original PBA is one that is marked with the old key ID and we know that it will be replaced as soon as its VBA gets rekeyed. However, the same inner node must be updated again and again in the context of the same rekeying but for other VBAs. This time, the original PBA of the CoW is already marked with the new key ID and we cannot detect its pseudo-reserved state as with the original PBA of the first update. These "dead" entries fill up the free tree and eventually render the CBE stuck.

As far as I can see, the CBE paper doesn't give a solution for this. One thing we could do is remember the logical address of the free tree entry of each reserved PBA in its Type-1 node in the VBD. This way, on CoW during rekeying, the free tree entry of the original PBA could also be updated efficiently and pseudo-reserved entries wouldn't emerge. Another solution would be to remember the VBD level for each reserved PBA in the free tree. As rekeying runs monotonic through the VBAs from 0 to highest, an inner node is updated once for each VBA that is greater or equal its first leaf VBA and less or equal its last leaf VBA. With the VBD level, the free tree knows the set of VBAs an inner node is updated for and is therefore able to detect pseudo-reserved inner node PBAs.

**Comment by m-stein** [comment]
The rekeying now finishes just fine (tested with a 32Mb device).

I could find another, more elegant solution for the above mentioned problem (I might document it here later) that also renders the original final clean up of the free tree unnecessary.

The next step will be to mix in other requests and test the "Online" in "Online Rekeying".

**Comment by m-stein** [comment]
Online Rekeying works fine on my cbe/cbe_20.05 and genode/cbe_20.05 Branches (also when intermixing other requests). The branches also add a fix that makes rekeying consider differently sized snapshots (due to resizing).
